cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java

```diff
@@ -0,0 +1,361 @@
+    } catch (Exception e) {
+      OUTAGE_LOG.warn("Failed to sync in pipeline {}. Will be retried.", name, e);
+    }
+    return config.getCheckpointIntervalMillis();
+  }
+
+  /**
+   * A mutable implementation of {@link Checkpoint}.
+   */
+  private static final class MutableLogBufferCheckpoint extends Checkpoint<LogBufferFileOffset> {
+    private MutableLogBufferFileOffset offset;
+    private long maxEventTs;
+
+    MutableLogBufferCheckpoint(long fileId, long filePos, long maxEventTs) {
+      super(new LogBufferFileOffset(fileId, filePos), maxEventTs);
+      this.offset = new MutableLogBufferFileOffset(fileId, filePos);
+      this.maxEventTs = maxEventTs;
+    }
+
+    @Override
+    public MutableLogBufferFileOffset getOffset() {
+      return offset;
+    }
+
+    @Override
+    public long getMaxEventTime() {
+      return maxEventTs;
+    }
+
+    void setMaxEventTs(long maxEventTs) {
+      this.maxEventTs = maxEventTs;
+    }
+
+    @Override
+    public String toString() {
+      return "MutableLogBufferCheckpoint{" +
+        "offset=" + offset +
+        ", maxEventTs=" + maxEventTs +
+        '}';
+    }
+  }
+
+  /**
+   * A mutable implementation of {@link LogBufferFileOffset}.
+   */
+  private static final class MutableLogBufferFileOffset extends LogBufferFileOffset {
+    private long fileId;
+    private long filePos;
+
+    MutableLogBufferFileOffset(long fileId, long filePos) {
+      super(fileId, filePos);
+      this.fileId = fileId;
+      this.filePos = filePos;
+    }
+
+    @Override
+    public long getFileId() {
+      return fileId;
+    }
+
+    @Override
+    public long getFilePos() {
+      return filePos;
+    }
+
+    void setFileId(long fileId) {
+      this.fileId = fileId;
+    }
+
+    void setFilePos(long filePos) {
+      this.filePos = filePos;
+    }
+
+    @Override
+    public String toString() {
+      return "MutableLogBufferFileOffset{" +
+        "fileId=" + fileId +
+        ", filePos=" + filePos +
+        '}';
+    }
+  }
+
+  /**
+   * Iterator to transform LogBufferEvent to ProcessorEvent.
+   */
+  private final class LogFileOffsetTransformIterator extends AbstractIterator<ProcessorEvent<LogBufferFileOffset>> {
+    private BlockingQueue<LogBufferEvent> queue;
+    private LogBufferEvent prevLogEvent;
+    int count = 0;
+
+    LogFileOffsetTransformIterator(BlockingQueue<LogBufferEvent> queue) {
+      this.queue = queue;
+    }
+
+    @Override
+    protected ProcessorEvent<LogBufferFileOffset> computeNext() {
+      // limit total number of events removed from incomingEventQueue to avoid infinite reading
+      if (count >= config.getBatchSize()) {
+        // if previous event was already processed, that should be removed.
+        removeProcessedEvent();
+        return endOfData();
+      }
+
+      if (queue.peek() == null) {
+        return endOfData();
+      }
+
+      // Remove already processed event from the queue
+      removeProcessedEvent();
```

Comment:
Not sure I understand the logic in this class. It seems like it would be simpler to remove prevLogEvent, poll here, increment, and return a new ProcessorEvent. How come we need to remove, then peek?

Reply:
The time event queue processor may not process the event if underlying queue is already full https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/queue/TimeEventQueueProcessor.java#L82 
When `iterator.hasNext` is called from time event queue processor, that call will indicate that  previous event has been processed. So we can safely remove `prevLogEvent` and then peek next event for further processing.

Reply:
How is that related to how this is implemented? Is the queue being read from somewhere else? Can't this be implemented like:

```
if (count >= batchSize) {
  return endOfData();
}

LogBufferEvent event = queue.poll();
if (event == null) {
  return endOfData();
}
count++;
return new ProcessorEvent<>(...);
```

Reply:
Ok, had an offline discussion with @albertshau about this. Since the AbstractIterator makes the implementation complicated, using `Iterator` is much simpler. In that case, iterator does not keep to keep track of `prevEvent`.